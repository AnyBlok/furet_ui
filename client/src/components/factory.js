import Vue from 'vue';

const components = {};
const componentNames = [];

export const defineComponent = (elementName, declaration) => {
  let klass;
  if (elementName in components) {
    klass = components[elementName];
  } else {
    klass = {
      template: undefined,
      extend: [],
      prototypes: [],
    };
    components[elementName] = klass;
    componentNames.push(elementName);
  }
  if (declaration.template !== undefined) klass.template = declaration.template;
  if (declaration.prototype !== undefined) {
    klass.prototypes.push(declaration.prototype);
  }
  if (declaration.extend !== undefined) {
    declaration.extend.forEach((extend) => {
      klass.extend.push(extend);
    });
  }
};

window.defineComponent = defineComponent;

const getPrototypeFor = (elementName) => {
  const component = components[elementName];
  const prototypes = [];
  component.extend.forEach((otherElementName) => {
    prototypes.push(...getPrototypeFor(otherElementName));
  });
  prototypes.push(...component.prototypes);
  return prototypes;
};

export const createComponents = () => {
  componentNames.forEach((elementName) => {
    const component = components[elementName];
    const mixins = getPrototypeFor(elementName);
    Vue.component(elementName, (resolve) => {
      resolve({
        template: component.template,
        mixins,
      });
    });
  });
};
