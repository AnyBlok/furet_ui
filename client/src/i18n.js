import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

export const i18nConf = {
  locale: 'en',
  fallbackLocale: 'en',
  messages: {
    en: {
    },
  },
};
export const i18n = new VueI18n(i18nConf);
