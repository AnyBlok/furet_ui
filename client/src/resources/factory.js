import _ from 'underscore';

const resources = {};
export const resourceNames = [];

export const removeResource = (resourceName) => {
  const index = resourceNames.indexOf(resourceName);
  if (index !== -1) {
    delete resources[resourceName];
    resourceNames.splice(index, 1);
  }
};

export const defineResource = (resourceName, declaration) => {
  let klass;
  if (resourceName in resources) {
    klass = resources[resourceName];
  } else {
    klass = {
      children: [],
      name: resourceName,
      path: resourceName,
      templateName: resourceName,
      props: [],
      mustBeAuthenticated: false,
    };
    resources[resourceName] = klass;
    resourceNames.push(resourceName);
  }
  if (declaration.path !== undefined) klass.path = declaration.path;
  if (declaration.templateName !== undefined) klass.templateName = declaration.templateName;
  if (declaration.props !== undefined) klass.props = declaration.props;
  if (declaration.mustBeAuthenticated !== undefined) {
    klass.mustBeAuthenticated = declaration.mustBeAuthenticated;
  }
  if (declaration.children !== undefined) {
    declaration.children.forEach((child) => {
      let childDefinition = _.find(klass.children, c => c.name === child.name);
      if (childDefinition === undefined) {
        childDefinition = {
          name: child.name,
          completeName: `${resourceName}_${child.name}`,
          templateName: child.name,
          path: child.name,
          props: [],
          mustBeAuthenticated: false,
        };
        klass.children.push(childDefinition);
      }
      if (child.path !== undefined) childDefinition.path = child.path;
      if (child.templateName !== undefined) childDefinition.templateName = child.templateName;
      if (child.props !== undefined) childDefinition.props.push(...child.props);
      if (child.mustBeAuthenticated !== undefined) {
        childDefinition.mustBeAuthenticated = child.mustBeAuthenticated;
      }
    });
  }
};

window.defineResource = defineResource;

const createRoute = (resource, defaultTemplateProps) => {
  let path = `/${resource.path}`;
  let templateProps = defaultTemplateProps || '';
  if (resource.path === resource.name) {
    if (resource.props.length !== 0) {
      resource.props.forEach((prop) => {
        path += `/:${prop}`;
        templateProps += ` v-bind:${prop}="${prop}"`;
      });
    }
  } else if (resource.props.length !== 0) {
    resource.props.forEach((prop) => {
      templateProps += ` v-bind:${prop}="${prop}"`;
    });
  }
  const route = {
    name: resource.completeName || resource.name,
    path,
    props: templateProps !== '',
    meta: { requiresAuth: resource.mustBeAuthenticated },
    component: { props: resource.props },
  };
  if (resource.children === undefined || resource.children.length === 0) {
    route.component.template = `<${resource.templateName}${templateProps} />`;
  } else {
    route.children = [];
    if (resource.templateName === resource.name) {
      route.component.template = '<router-view />';
    } else {
      route.component.template = `<${resource.templateName}${templateProps} />`;
    }
    resource.children.forEach((child) => {
      route.children.push(createRoute(child, templateProps));
    });
  }
  return route;
};

export const getRoutes = () => {
  const routes = [];
  resourceNames.forEach((resourceName) => {
    const resource = resources[resourceName];
    routes.push(createRoute(resource));
  });
  return routes;
};
