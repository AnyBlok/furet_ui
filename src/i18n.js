/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';
import VueI18n from 'vue-i18n';
Vue.use(VueI18n);


export const i18nConf = {
    locale: 'en',
    fallbackLocale: 'en',
    messages: {
        en: {
            menus: {
                close: 'Close',
                search: 'Search ...',
            },
            views: {
                unknown: {
                    title: 'The wanted view "{name}" is unknown',
                    message: "Report this message to the administrator",
                },
                common: {
                    create: 'Create',
                    save: 'Save',
                    edit: 'Edit',
                    cancel: 'Cancel',
                    delete: 'Delete',
                    close: 'Close',
                    actions: 'Actions',
                    more: 'More',
                },
                clients: {
                    login: {
                        button: 'Log in',
                    },
                },
            },
            fields: {
                common: {
                    required: 'this field is required',
                },
                json: {
                    invalid: 'Invalid JSON format',
                },
                file: {
                    upload: 'Upload',
                    download: 'Download',
                    delete: 'Delete',
                },
            },
        },
    },
};
export const i18n = new VueI18n(i18nConf);
