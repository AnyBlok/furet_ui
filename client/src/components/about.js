import { defineComponent } from './factory';

defineComponent('about', {
  template: `
    <section class="section">
      <div class="container has-text-centered">
        <ul>
          <li><strong>version</strong> 0.0.0</li>
        </ul>
      </div>
    </section>
  `,
});
