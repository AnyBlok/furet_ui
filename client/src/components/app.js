import { defineComponent } from './factory';

defineComponent('app', {
  template: `
    <div id="furet-ui-app">
      <furet-ui-appbar />
      <notifications />
      <router-view></router-view>
      <furet-ui-footer />
    </div>`,
});

defineComponent('furet-ui-appbar', {
  template: `
    <header class="hero is-dark">
      <furet-ui-appbar-header />
      <furet-ui-appbar-body />
      <furet-ui-appbar-footer />
    </header>
  `,
});

defineComponent('furet-ui-appbar-header', {
  template: `
    <div class="hero-head">
      <nav class="navbar is-dark" role="navigation">
        <div class="container">
          <div class="navbar-brand">
            <furet-ui-appbar-header-brand />
            <a 
              role="button" 
              v-bind:class="['navbar-burger', 'burger', isOpen ? 'is-active' : '']" 
              aria-label="menu" 
              aria-expanded="false" 
              data-target="navbarUserMenu" 
              v-on:click="isOpen = !isOpen"
            >
              <span aria-hidden="true"></span>
              <span aria-hidden="true"></span>
              <span aria-hidden="true"></span>
            </a>
          </div>
          <div id="navbarUserMenu" v-bind:class="['navbar-menu', isOpen ? 'is-active' : '']">
            <furet-ui-appbar-spaces-menu />
            <furet-ui-appbar-user-menu />
          </div>
        </div>
      </nav>
    </div>
  `,
  prototype: {
    data() {
      return {
        isOpen: false,
      };
    },
  },
});

defineComponent('furet-ui-appbar-header-brand', {
  template: `
    <img class="image is-64x64" v-bind:src="logo" alt="Logo">
  `,
  extend: ['mixin-logo'],
});

defineComponent('furet-ui-appbar-body', {
  template: `
    <div class="hero-body">
      <div class="container has-text-centered">
        <h1 class="title">
          Furet UI
        </h1>
        <p class="subtitle">
          Web Client for any backend serveur
        </p>
      </div>
    </div>
  `,
});

defineComponent('furet-ui-appbar-footer', {
  template: `
    <div class="hero-foot" v-if="hasSpaceMenus">
      <furet-ui-appbar-space-menus />
    </div>
  `,
  prototype: {
    computed: {
      hasSpaceMenus() {
        return this.$store.state.menus.spaceMenus.length !== 0;
      },
    },
  },
});

defineComponent('furet-ui-appbar-user-menu', {
  prototype: {
    render(createElement) {
      const menus = [];
      this.menus.forEach((menu) => {
        menus.push(createElement(
          menu.component,
          {
            class: menu.class,
            style: menu.style,
            attrs: menu.attrs,
            props: menu.props,
            domProps: menu.domProps,
            on: menu.on,
            nativeOn: menu.nativeOn,
          },
          [menu.label]));
      });
      return createElement('div', { class: 'navbar-end' }, menus);
    },
    computed: {
      menus() {
        return this.$store.state.menus.user;
      },
    },
  },
});

defineComponent('furet-ui-appbar-router-link-goto', {
  prototype: {
    props: ['label', 'to'],
    methods: {
      goTo() {
        // TODO clear data
        this.$router.push(this.to);
      },
    },
  },
});

defineComponent('furet-ui-appbar-head-router-link', {
  template: `
    <a class="navbar-item" v-on:click="goTo">
      {{ label }}
    </a>
  `,
  extend: ['furet-ui-appbar-router-link-goto'],
});

defineComponent('furet-ui-appbar-head-router-link-button', {
  template: `
    <span class="navbar-item">
      <a class="button is-dark is-inverted is-fullwidth" v-on:click="goTo">
        <span class="icon" v-if="icon">
          <font-awesome-icon v-bind:icon="icon" />
        </span>
        <span>{{ label }}</span>
      </a>
    </span>
  `,
  extend: ['furet-ui-appbar-router-link-goto'],
  prototype: {
    props: ['icon'],
  },
});

defineComponent('furet-ui-appbar-spaces-menu', {
  prototype: {
    render(createElement) {
      const menus = [];
      this.menus.forEach((menu) => {
        menus.push(createElement(
          menu.component,
          {
            class: menu.class,
            style: menu.style,
            attrs: menu.attrs,
            props: menu.props,
            domProps: menu.domProps,
            on: menu.on,
            nativeOn: menu.nativeOn,
          },
          [menu.label]));
      });
      return createElement('div', { class: 'navbar-start' }, menus);
    },
    computed: {
      menus() {
        return this.$store.state.menus.spaces;
      },
    },
  },
});

defineComponent('furet-ui-appbar-space-menus', {
  prototype: {
    render(createElement) {
      const menus = [];
      this.menus.forEach((menu) => {
        menus.push(createElement('li', [
          createElement(
            menu.component,
            {
              class: menu.class,
              style: menu.style,
              attrs: menu.attrs,
              props: menu.props,
              domProps: menu.domProps,
              on: menu.on,
              nativeOn: menu.nativeOn,
            },
            [menu.label]),
        ]));
      });
      return createElement('nav', { class: 'tabs is-boxed' }, [
        createElement('div', { class: 'container' }, [
          createElement('ul', [menus]),
        ]),
      ]);
    },
    computed: {
      menus() {
        return this.$store.state.menus.spaceMenus;
      },
    },
  },
});

defineComponent('furet-ui-appbar-foot-router-link', {
  template: `
    <a v-on:click="goTo">
      {{ label }}
    </a>
  `,
  extend: ['furet-ui-appbar-router-link-goto'],
});

defineComponent('furet-ui-appbar-foot-router-link-button', {
  template: `
    <a class="button is-dark is-inverted" v-on:click="goTo">
      <span class="icon" v-if="icon">
        <font-awesome-icon v-bind:icon="icon" />
      </span>
      <span>{{ label }}</span>
    </a>
  `,
  extend: ['furet-ui-appbar-router-link-goto'],
  prototype: {
    props: ['icon'],
  },
});

defineComponent('furet-ui-footer', {
  template: `
    <footer class="footer">
      <div class="content has-text-centered is-success">
        <p>
          <strong>Furet UI</strong> by <a href="https://github.com/jssuzanne">Jean-SÃ©bastien Suzanne</a>. 
          The source code is licensed <a href="http://opensource.org/licenses/MPL-2.0">MPL-2.0</a> and 
          is available on <a href="https://github.com/AnyBlok/furet_ui">Github</a>.
        </p>
      </div>
    </footer>
  `,
});
