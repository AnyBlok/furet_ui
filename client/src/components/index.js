import './mixins';
import './app';
import './homepage';
import './login';
import './logout';
import './about';
import './list';

export { createComponents, default } from './factory';
