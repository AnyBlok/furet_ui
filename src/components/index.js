import './mixins';
import './app';
import './homepage';
import './login';
import './logout';
import './pages';
import './fields';

export { createComponents, default } from './factory';
