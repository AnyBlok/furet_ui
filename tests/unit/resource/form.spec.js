import { mount } from "@vue/test-utils";
import { getComponentPrototype } from '@/components/factory';
import { i18n } from "@/i18n";
// import { eval_counter } from "@/components/resource/helper";
import axios from 'axios';

const localVue = global.localVue;
const router = global.router;

const mocki18n = jest.spyOn(i18n, "_t");
mocki18n.mockResolvedValue('');

const mock = jest.spyOn(axios, "get");
mock.mockResolvedValue({data: [], headers: {'x-total-records': 0}});

const mockAxios = jest.spyOn(axios, "post");
mockAxios.mockImplementationOnce(() =>
  Promise.resolve({ data: "redirect/uri" })
);

const store = global.store

const getEntry = (model, pk) => {
  model;
  pk;
  return {title: 'Entry 1', color: 'red'};
}

const getNewEntry = (model, uuid) => {
  model;
  uuid;
  return {title: 'New Entry 1', color: 'red'};
}

const updateChangeState = () => {}

describe('furet-ui-resource-form component', () => {

  store.commit('UPDATE_RESOURCES', {'definitions': [{
    'id': 1, 
    'type': 'form', 
    'model': 'Model.1', 
    'header_template': `
      <nav class="level">
        <div class="level-left">
          <div class="level-item content">
            <h1 class="is-large">{{ data.title }}</h1>
          </div>
        </div>
      </nav>
    `,
    'body_template': `
      <div 
        id="id" 
        key="key"
      >
        <div class="columns">
          <div class="column">
            <furet-ui-field 
              name="title" 
              v-bind:config="{name: 'title', type: 'string', label: 'Title'}" 
              v-bind:resource="resource"
              v-bind:data="data"
            >
            </furet-ui-field>
          </div>
          <div class="column">
            <furet-ui-field 
              name="color" 
              v-bind:config="{name: 'color', type: 'string', label: 'Color'}" 
              v-bind:resource="resource"
              v-bind:data="data"
            >
            </furet-ui-field>
          </div>
        </div>
      </div>
    `,
    'footer_template': `
      <nav class="level">
        <div class="level-left">
          <div class="level-item content">
            <h1 class="is-large">{{ data.title }}</h1>
          </div>
        </div>
      </nav>
    `,
    'fields': ['id', 'title', 'color']
  }]});

  beforeEach(() => {
    mock.mockClear();
    mockAxios.mockClear();
  });

  const wrapper = mount(getComponentPrototype("furet-ui-resource-form"), {
    store,
    localVue,
    router,
    i18n,
    propsData: {
      id: 1,
      manager: {query: {pks: JSON.stringify({id: 1})}},
    },
    provide: {
      getEntry,
      getNewEntry,
      updateChangeState,
    }
  });
  it('furet-ui-resource-form: snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
  it('furet-ui-resource-form: getBreadcrumbInfo', () => {
    const info = wrapper.vm.getBreadcrumbInfo()
    expect(info).toStrictEqual({icon: "newspaper", label: ""})
  });
  it('refresh callback', () => {
    wrapper.vm.registryRefreshCallback(() => {});
  });
  it('furet-ui-resource-form: readonly.', () => {
    expect(wrapper.vm.isReadonly()).toBe(wrapper.vm.readonly);
  });
  it('furet-ui-resource-form: loadAsyncData.', () => {
    wrapper.vm.loadAsyncData()
    expect(mock).toBeCalledWith(
      "/furet-ui/resource/1/crud", 
      {
        params: {
          "context[fields]": "id,title,color",
          "context[model]": "Model.1",
          "filter[id][eq]": 1,
        }
      }
    );
  });
  it('furet-ui-resource-form: refresh.', () => {
    wrapper.vm.refresh()
    expect(mock).toBeCalledWith(
      "/furet-ui/resource/1/crud", 
      {
        params: {
          "context[fields]": "id,title,color",
          "context[model]": "Model.1",
          "filter[id][eq]": 1,
        }
      }
    );
  });
});

describe('furet-ui-form-button component', () => {
  it('snapshot isReadonly', () => {
    const wrapper = mount(getComponentPrototype("furet-ui-form-button"), {
      store,
      localVue,
      router,
      i18n,
      propsData: {
        resource: {
        },
        data: {},
        config: {
          label: 'Test',
          icon: 'th',
        },
      },
      provide: {
        partIsReadonly: () => {},
        currentResource: {
          uuid: null,
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot not isReadonly', () => {
    const wrapper = mount(getComponentPrototype("furet-ui-form-button"), {
      store,
      localVue,
      router,
      i18n,
      propsData: {
        resource: {
          readonly: true,
        },
        data: {},
        config: {
          readonly: false,
          label: 'Test',
          icon: 'th',
        },
      },
      provide: {
        partIsReadonly: () => true,
        currentResource: {
          uuid: null,
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot with class', () => {
    const wrapper = mount(getComponentPrototype("furet-ui-form-button"), {
      store,
      localVue,
      router,
      i18n,
      propsData: {
        resource: {
          readonly: true,
        },
        data: {},
        config: {
          readonly: false,
          label: 'Test',
          icon: 'th',
          class: ['c1', 'c2']
        },
      },
      provide: {
        partIsReadonly: () => true,
        currentResource: {
          uuid: null,
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot isHidden', () => {
    const wrapper = mount(getComponentPrototype("furet-ui-form-button"), {
      store,
      localVue,
      router,
      i18n,
      propsData: {
        resource: {
        },
        data: {},
        config: {
          hidden: true,
          label: 'Test',
          icon: 'th',
        },
      },
      provide: {
        partIsReadonly: () => {},
        currentResource: {
          uuid: null,
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot server_call', () => {
    const wrapper = mount(getComponentPrototype("furet-ui-form-button"), {
      store,
      localVue,
      router,
      i18n,
      propsData: {
        resource: {
          pks: {id: 1},
          model: 'Model.1',
          id: 1
        },
        data: {},
        config: {
          label: 'Test',
          icon: 'th',
          call: 'test',
        },
      },
      provide: {
        partIsReadonly: () => {},
        currentResource: {
          uuid: null,
        }
      }
    });
    wrapper.vm.server_call()
    expect(mockAxios).toBeCalledWith(
      "/furet-ui/resource/1/model/Model.1/call/test", 
      {pks: {id: 1}}
    );
  });
});
