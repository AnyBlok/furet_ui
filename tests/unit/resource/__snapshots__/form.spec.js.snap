// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`furet-ui-form-button component snapshot isHidden 1`] = `<!---->`;

exports[`furet-ui-form-button component snapshot isReadonly 1`] = `
<button
  class="button is-primary is-outlined"
  disabled="disabled"
>
  <span
    class="icon"
  >
    <span
      class="icon"
    >
      <i
        class="fa fa-th fa-lg"
      />
    </span>
  </span>
   
  <span>
    Test
  </span>
</button>
`;

exports[`furet-ui-form-button component snapshot not isReadonly 1`] = `
<button
  class="button is-primary is-outlined"
>
  <span
    class="icon"
  >
    <span
      class="icon"
    >
      <i
        class="fa fa-th fa-lg"
      />
    </span>
  </span>
   
  <span>
    Test
  </span>
</button>
`;

exports[`furet-ui-form-button component snapshot with class 1`] = `
<button
  class="button c1 c2"
>
  <span
    class="icon"
  >
    <span
      class="icon"
    >
      <i
        class="fa fa-th fa-lg"
      />
    </span>
  </span>
   
  <span>
    Test
  </span>
</button>
`;

exports[`furet-ui-resource-form component furet-ui-resource-form: snapshot 1`] = `
<section>
  <div
    class="loading-overlay is-active is-full-page"
  >
    <div
      class="loading-background"
    />
     
    <div
      class="loading-icon"
    />
  </div>
   
  <header
    id="header_page"
    name="furet-ui-page"
  >
    <nav
      class="level"
    >
      <div
        class="level-left"
      >
        <div
          class="level-item content"
        >
          <h1
            class="is-large"
          >
            Entry 1
          </h1>
        </div>
      </div>
    </nav>
     
    <div
      class="buttons is-grouped is-centered"
    >
      <a
        class="button is-primary is-outlined"
      >
        <span
          class="icon"
        >
          <span
            class="icon"
          >
            <i
              class="fa fa-arrow-left fa-lg"
            />
          </span>
        </span>
         
        <span>
          [object Promise]
        </span>
      </a>
       
      <!---->
       
      <!---->
       
      <!---->
       
      <!---->
       
      <!---->
       
      <!---->
       
    </div>
  </header>
   
  <div
    class="section"
  >
    <div
      id="id"
    >
      <div
        class="columns"
      >
        <div
          class="column"
        >
          <span
            class="is-primary is-top is-medium b-tooltip"
            data-label=""
            name="title"
            style="transition-delay: 0ms; width: 100%;"
          >
            <div
              class="field"
              style="width: inherit;"
            >
              <label
                class="label"
              >
                Title
              </label>
               
              <div>
                <span>
                  Entry 1
                </span>
              </div>
               
              <!---->
            </div>
          </span>
        </div>
         
        <div
          class="column"
        >
          <span
            class="is-primary is-top is-medium b-tooltip"
            data-label=""
            name="color"
            style="transition-delay: 0ms; width: 100%;"
          >
            <div
              class="field"
              style="width: inherit;"
            >
              <label
                class="label"
              >
                Color
              </label>
               
              <div>
                <span>
                  red
                </span>
              </div>
               
              <!---->
            </div>
          </span>
        </div>
      </div>
    </div>
  </div>
   
  <nav
    class="level"
  >
    <div
      class="level-left"
    >
      <div
        class="level-item content"
      >
        <h1
          class="is-large"
        >
          Entry 1
        </h1>
      </div>
    </div>
  </nav>
</section>
`;
